### PR TITLE
feat: Production Plan: make the item_code filter into a 'like' search rather than a link field

### PR DESCRIPTION
**Pull request by nameduser0**
Allows you to add multiple items with similar item_codes to the same production plan rather than the strict single item according to a link field

Consider a realistic scenario: a clothes printing shop is much more likely to want to make all t-shirts, or all medium t-shirts, or all white t-shirts, or all t-shirt sizes with a single design (new behaviour) than they are to make a single t-shirt design, in a single size, in a single colour (existing behaviour).

no-docs

### CODE REVIEW

**Comment by nameduser0** [comment]
Won't-fix old format SQL

File: erpnext/manufacturing/doctype/production_plan/production_plan.py
Comment (ankush) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]:
```suggestion
		pending_mr = frappe.db.sql(""" 
```

This is becoming part of the query string. I've removed the rule that's causing unnecessary warnings. So you can undo this change.

**Comment by ankush** [comment]
This looks neat 👍  Especially because planning is usually not done on item level it makes sense to use "LIKE" filtering... will do some manual testing before merging.

**Comment by nameduser0** [comment]
@ankush Thanks. It's not the best solution. I would have liked to change the buttons to be additive instead of clear table and fill from scratch, but then I would need to implement checks to see if the lines were already added. This quick fix is the next best thing.

File: erpnext/manufacturing/doctype/production_plan/production_plan.json
Comment (ankush) [inline erpnext/manufacturing/doctype/production_plan/production_plan.json]: Replace "string" with text, string is programmer-talk. Or remove the full description. We usually don't like writing descriptions, it should be obvious wherever possible.

File: erpnext/manufacturing/doctype/production_plan/production_plan.py
Comment (ankush) [inline erpnext/manufacturing/doctype/production_plan/production_plan.py]: when `self.item_code` is not entered by the user, this will give a string concatenation error. (Same with SO query down below)

**Review by ankush — CHANGES_REQUESTED** [review]

**Comment by ankush** [comment]
btw we were thinking of removing all these filter fields and adding modal dialog (already present in many other documents) Might take some time for this though. 

this is from picklist:
Screenshot 2021-07-30 at 3 30 20 PM

File: erpnext/manufacturing/doctype/production_plan/production_plan.json
Comment (nameduser0) [reply erpnext/manufacturing/doctype/production_plan/production_plan.json]: I'll change the word but I think this is the wrong policy on descriptions.

This is an extremely powerful and easy self-documentation opportunity. ERP systems are very much **not** straightforward to pick up, each field is rarely obvious and often has knock-on implications in many related documents and fields. This one field for example doesn't just get used to filter Sales Orders, but also the Items that are added to the PP. This is not obvious.

My opinion (who has spent several years with ERP system users) is that **every** field or section should have a description, unless it truly is obvious like `Date of Birth` kind of stuff. A user shouldn't need to refer to the documentation to make sense of the form, which incidentally rarely fully describes the features anyway.

Another very much not obvious field is `Include Exploded Items`. I had to read the code before I understood what that did. Who wants a production plan that doesn't `Include Exploded Items` ? As it turns out, it's just a broken, badly named feature that I had to fix which I renamed `Make From Scratch`.

**Comment by nameduser0** [comment]
> btw we were thinking of removing all these filter fields and adding modal dialog (already present in many other documents) Might take some time for this though.

Makes sense if it adds functionality, like a full set of custom field filters. But in general I don't agree many dialogs are the answer, they break the user experience, like the Warehouse dialog on `Get Raw Materials For Production`. This should be on-page.

Dialogs are also a bad idea if they force the user to enter the same thing twice, like filtering for Sales Orders containing an Item, then filtering again for Items. Or if they lose work that takes time. Adding to the existing fields is easy because it's data on the page that is saved, so easy to edit. On a dialog the filters would be lost.

Another example of pointless feature creep is the `awesomplete` js that moves the page around every time you complete a field. This is extremely annoying. Users like things to be where they left them, not constantly searching for them because they moved or the section was auto-collapsed. Imagine someone moving all your stuff around your room every day without asking and with no way to stop it. You would quickly move.

In general, your time would be better spent making existing functionality 100% bug-free and usable than continuous feature-creep and refactoring. ERPNext is losing customers daily.

**Comment by nameduser0** [comment]
[https://erpnext.org/contribute](https://erpnext.org/contribute) 😆

File: erpnext/manufacturing/doctype/production_plan/production_plan.py
Comment (ankush) [reply erpnext/manufacturing/doctype/production_plan/production_plan.py]: I don't think the change you did in last commit will solve this. 

F-strings will convert `None` to string literal `'None'` and then search for item code like `%None%`

Comment (nameduser0) [reply erpnext/manufacturing/doctype/production_plan/production_plan.py]: In all instances the additional filter conditions are only added if item_code is truthy

**Comment by ankush** [comment]
LGTM. Just resolve conflicts or allow maintainer edits.

**Comment by nameduser0** [comment]
> LGTM. Just resolve conflicts or allow maintainer edits.

@ankush See comments on #25845. conflicts are unresolvable due to that breaking change.

**Comment by nameduser0** [comment]
@ankush Your last comment on the other thread is also incorrect. A BOM captures both the raw materials (subitems) and process (operations). So as I stated, if items have the same BOM, they are by definition the same item. And if this is how the system is being used, it is simply abuse.

**Comment by ankush** [comment]
Not everyone captures all the processes in BOM. We are software developers, we don't enforce good systems to end-users.  🤷 

If you want this strict check then add a setting in manufacturing settings to disallow selling/fetching BOM template... even make it the default behaviour.

**Pull request closed without merging**
